### ISSUE TITLE
perf: AVX-512 VNNI histogram construction for the quantized-gradient path

### ISSUE THREAD

**Issue by kyo219** [kyo219/LightGBM-MoE#18]
## Background

The dominant hot loop on the quantized-gradient path is `ConstructHistogramIntInner` in `src/io/dense_bin.hpp:175-222`:

\`\`\`cpp
const auto ti = static_cast<uint32_t>(data(idx));
const int16_t gradient_16 = gradients_ptr[i];
const PACKED_HIST_T gradient_packed = ...;  // pack int8(grad)|int8(hess)
out_ptr[ti] += gradient_packed;             // scatter accumulate
\`\`\`

This is a classic gather/scatter accumulate pattern that the compiler **cannot autovec** — `ti` is data-dependent, and the same bin can be written multiple times within a single SIMD lane (RAW hazard).

Adding `-march=native -mtune=native` via the new `USE_NATIVE_ARCH` CMake option (commit `e300d07d`) moved overall training time by ~0% because of exactly this pattern: the loop is memory-bound, not compute-bound, so wider scalar registers don't help.

## Goal

A manual AVX-512 VNNI implementation that turns this into a true int8 × uint8 → int32 dot product per histogram bin, cutting the histogram phase by 2-4×.

## Sketch

For each block of (e.g.) 64 samples:

1. Load 64 sample bin indices into a `__m512i` (uint8 lane).
2. Detect collisions with `_mm512_conflict_epi32` — colliding lanes serialize, non-colliding accumulate in parallel.
3. For non-colliding lanes, express the update as `vpdpbusd(packed_grad_lane, one_hot_bin, accumulator_lane)`.
4. Pair with **per-thread histogram tiling**: each OMP thread accumulates into a private aligned histogram, then a final reduction sums them. This breaks the memory-bandwidth bottleneck that is currently masking any compute improvement.

## Files

- `src/io/dense_bin.hpp` — new `#ifdef __AVX512VNNI__` specialization of `ConstructHistogramIntInner` next to the existing scalar template.
- Possibly `src/treelearner/feature_histogram.hpp` for matching reduction.
- `CMakeLists.txt` — add an AVX-512 detection check (compile-time `__AVX512VNNI__` guard, runtime `__builtin_cpu_supports(\"avx512vnni\")` fallback for distributed binaries).

## Pre-conditions

- `USE_NATIVE_ARCH=ON` (the build option already exists).
- AVX-512 VNNI-capable CPU: Zen4+ (Ryzen 7000+, EPYC 9004+), Intel Ice Lake+, or Sapphire Rapids+.
- Quantized gradient path active (`use_quantized_grad=true`).

## Acceptance criteria

- `examples/perf_bench.py` shows ≥2× speedup on `gbdt/quant16` and `moe/quant16` configs versus the scalar version, holding RMSE within 0.01.
- Falls back gracefully on CPUs without VNNI (no segfault, scalar path used).
- Bench numbers added to `bench_results/` for the new build.

## Scope / risks

- Realistic implementation effort: 2-3 weeks including correctness testing and the runtime CPU-feature dispatch.
- Histogram tiling changes the reduction semantics slightly; need careful equivalence tests vs the scalar reference.
- Inline TODO marker is already in `src/io/dense_bin.hpp` (commit `1183c543`) so a future contributor can pick this up cold.

## References

- Phase 3 deferral notes on branch `feat/int8-perf` (2026-04-30 session)
- Inline sketch at `src/io/dense_bin.hpp` above `ConstructHistogramIntInner`
- Intel intrinsics: `_mm512_dpbusd_epi32`, `_mm512_conflict_epi32`
